druid/src/widget/textbox/editable_text.rs

```diff
@@ -0,0 +1,212 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Traits for text editing and a basic String implementation.
+
+use crate::unicode_segmentation::GraphemeCursor;
+
+pub trait EditableText {
+    // pub fn edit<T: IntervalBounds>(&mut self, iv: T, new: String);
+
+    // pub fn slice<T IV>(&self, iv: T) -> String;
+
+    // pub fn is_codepoint_boundary(&self, offset: usize) -> bool;
+
+    // fn prev_codepoint_offset(&self, offset: usize) -> Option<usize>;
+
+    // fn next_codepoint_offset(&self, offset: usize) -> Option<usize>;
+
+    // pub fn at_or_next_codepoint_boundary(&self, offset: usize) -> Option<usize>;
+
+    // pub fn at_or_prev_codepoint_boundary(&self, offset: usize) -> Option<usize>;
+
+    fn prev_grapheme_offset(&self, offset: usize) -> Option<usize>;
+
+    fn next_grapheme_offset(&self, offset: usize) -> Option<usize>;
+}
+
+impl EditableText for &str {
+    /// Gets the next grapheme from the given index.
+    fn next_grapheme_offset(&self, from: usize) -> Option<usize> {
+        let mut c = GraphemeCursor::new(from, self.len(), true);
+        c.next_boundary(self, 0).unwrap()
+    }
+
+    /// Gets the previous grapheme from the given index.
+    fn prev_grapheme_offset(&self, from: usize) -> Option<usize> {
+        let mut c = GraphemeCursor::new(from, self.len(), true);
+        c.prev_boundary(self, 0).unwrap()
+    }
+}
+
+impl EditableText for &mut String {
+    /// Gets the next grapheme from the given index.
+    fn next_grapheme_offset(&self, from: usize) -> Option<usize> {
+        let mut c = GraphemeCursor::new(from, self.len(), true);
+        c.next_boundary(self, 0).unwrap()
+    }
+
+    /// Gets the previous grapheme from the given index.
+    fn prev_grapheme_offset(&self, from: usize) -> Option<usize> {
+        let mut c = GraphemeCursor::new(from, self.len(), true);
+        c.prev_boundary(self, 0).unwrap()
+    }
+}
+
+impl EditableText for &String {
+    /// Gets the next grapheme from the given index.
+    fn next_grapheme_offset(&self, from: usize) -> Option<usize> {
+        let mut c = GraphemeCursor::new(from, self.len(), true);
+        c.next_boundary(self, 0).unwrap()
+    }
+
+    /// Gets the previous grapheme from the given index.
+    fn prev_grapheme_offset(&self, from: usize) -> Option<usize> {
+        let mut c = GraphemeCursor::new(from, self.len(), true);
+        c.prev_boundary(self, 0).unwrap()
+    }
+}
+
+pub trait EditableTextCursor<'a> {
```

Comment:
To check my understanding: each type that implements `EditableText` will have a corresponding cursor type?

In this case, we may want to reflect this relationship in the type system, where possible. In particular, the constructor for the cursor should take the type of the corresponding `EditableText`.

This is a bit tricky. One option would be to have the `EditableText` type by an associated type of `EditableTextCursor`, so you'd have something like

```rust
pub trait EditableTextCursor<'a> {
    type Text: EditableText;
    
    fn new(s: &'a Self::Text, position: usize) -> Self;

    // ...
}
```

but it might be simpler to go the other way, and have the cursor be the associated type of the text, e.g.

```rust
pub trait EditableText {
   type Cursor: EditableTextCursor<Self>;

   fn cursor(&self, position: usize) -> Self::Cursor;
}
```

Which of these makes the most sense probably comes down to how the API is used, which type gets passed around more, etc. What we want to be able to do is to remove references to concrete types like `str` & `String` where possible, and doing this requires making the types 'aware' of each other, somehow.

Reply:
I went with this latter version, which makes sense to me. But also: we might not even need the separate cursor type?